django/core/mail/message.py

```diff
@@ -71,56 +72,57 @@ def forbid_multi_line_headers(name, val, encoding):
     return name, val
 
 
-def split_addr(addr, encoding):
-    """
-    Split the address into local part and domain and encode them.
-
-    When non-ascii characters are present in the local part, it must be
-    MIME-word encoded. The domain name must be idna-encoded if it contains
-    non-ascii characters.
-    """
-    if '@' in addr:
-        localpart, domain = addr.split('@', 1)
-        # Try to get the simplest encoding - ascii if possible so that
-        # to@example.com doesn't become =?utf-8?q?to?=@example.com. This
-        # makes unit testing a bit easier and more readable.
-        try:
-            localpart.encode('ascii')
-        except UnicodeEncodeError:
-            localpart = Header(localpart, encoding).encode()
-        domain = domain.encode('idna').decode('ascii')
-    else:
-        localpart = Header(addr, encoding).encode()
-        domain = ''
-    return (localpart, domain)
-
-
 def sanitize_address(addr, encoding):
     """
     Format a pair of (name, address) or an email address string.
     """
+    address = None
+
     if not isinstance(addr, tuple):
-        addr = parseaddr(addr)
-    nm, addr = addr
-    localpart, domain = None, None
+        addr = force_str(addr)
+        try:
+            # This will try either to parse 'Name <add@ress>' or 'add@ress'
+            # get_mailbox raises an IndexError when given "a@"
+            # and a HeaderParseError when given "@"
+            token, rest = parser.get_mailbox(addr)
+            if not isinstance(token, Mailbox):
+                raise ValueError
+        except (HeaderParseError, ValueError, IndexError):
+            raise ValueError("Invalid address '{}'".format(addr))
+
+        else:
+            # If we're here, there are still 2 cases, either we parsed the full
+            # input or not. If there's still a rest, we probably misparsed.
+            if rest:
+                raise ValueError(
+                    "Invalid adddress; only '{}' "
+                    "could be parsed from '{}'".format(token, addr))
+
+            nm = token.display_name or ""
+            localpart = token.local_part
+            domain = token.domain or ""
+
+    else:
+        nm, address = addr
+        localpart, domain = address.rsplit('@', 1)
```

Comment:
Can we add a test that is falling with `split()` instead `rsplit()`?

Reply:
The test mentionned in https://github.com/django/django/pull/11417#discussion_r293158223 fails with split with
```
- =?utf-8?q?To_Example?= <to@other.com@example.com>
+ =?utf-8?q?To_Example?= <"to@other.com"@example.com>
```